### PR TITLE
[JAX] Change dtype of intermediate result aval of fused_topk_and_score_function_fwd to fp32

### PR DESCRIPTION
**Pull request by tdophung**
Fixed aval for intermediate results (softmaxed/sigmoided logits from fwd) to pass as residuals to `fused_topk_and_score_function_bwd` to CompType, which is currently hardcoded to fp32 in `transformerengine/common`. This prevents incorrect reading of this buffer when logits dtype used are not the same type as fp32. This was observed when integrating router to maxtext, and while grads flowing into `fused_topk_and_score_function_bwd` is clean of nans, the output from this kernel were nans.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Change dtype of intermediate result to CompType

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by tdophung** [comment]
Since @denera  mentioned that CompType will soon be configurable, instead of hard coded to fp32, I wonder if I should just export the CompType to `router.cpp` and set the aval dtype in `cpp_extensions/router.py` to that exported CompType instead of hardcoding like this to fp32. 

Wdyt @jberchtold-nvidia @phu0ngng ?

File: transformer_engine/jax/csrc/extensions/router.cpp [hunk 1]
```diff
@@ -41,7 +41,8 @@ Error_Type FusedTopkWithScoreFunctionForwardFFI(
   auto logits_tensor = TensorWrapper(logits, flat_shape, dtype);
   auto probs_tensor = TensorWrapper(probs, flat_shape, dtype);
   auto routing_map_tensor = TensorWrapper(routing_map, flat_shape, DType::kByte);
-  auto intermediate_tensor = TensorWrapper(intermediate, flat_shape, dtype);
+  // intermediate is always float32 (CompType) regardless of logits dtype.
+  auto intermediate_tensor = TensorWrapper(intermediate, flat_shape, DType::kFloat32);
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/csrc/extensions/router.cpp:51]: Can we add an `NVTE_CHECK` to assert the dtype of `intermediate_buf` is float32 just in case something every breaks on the Python abstract side?

**Comment by jberchtold-nvidia** [comment]
> Since @denera mentioned that CompType will soon be configurable, instead of hard coded to fp32, I wonder if I should just export the CompType to `router.cpp` and set the aval dtype in `cpp_extensions/router.py` to that exported CompType instead of hardcoding like this to fp32.
> 
> Wdyt @jberchtold-nvidia @phu0ngng ?

I have a slight preference to wait on this until the change has been made in TE common and have this PR just be the bugfix to make the intermediate buf always fp32.

If we make this change to have a generic CompType now before TE common is ready, we could get ahead on the boilerplate logic here. However, we'll still need asserts in Python to prevent users from passing a comp dtype other than fp32 since TE common as of now doesn't support it. Then once TE common supports compute dtypes other than fp32, we'll need a follow-up TE/JAX PR to update this assert to the new supported compute dtypes and add unit tests for them.

Given we'll need a follow-up PR here anyways once TE common supports other compute dtypes, I feel like it's easier to have this PR just be the bugfix and wait for the generalizing until TE common changes have merged.

Wdyt?

**Comment by tdophung** [comment]
> > Since @denera mentioned that CompType will soon be configurable, instead of hard coded to fp32, I wonder if I should just export the CompType to `router.cpp` and set the aval dtype in `cpp_extensions/router.py` to that exported CompType instead of hardcoding like this to fp32.
> > Wdyt @jberchtold-nvidia @phu0ngng ?
> 
> I have a slight preference to wait on this until the change has been made in TE common and have this PR just be the bugfix to make the intermediate buf always fp32.
> 
> If we make this change to have a generic CompType now before TE common is ready, we could get ahead on the boilerplate logic here. However, we'll still need asserts in Python to prevent users from passing a comp dtype other than fp32 since TE common as of now doesn't support it. Then once TE common supports compute dtypes other than fp32, we'll need a follow-up TE/JAX PR to update this assert to the new supported compute dtypes and add unit tests for them.
> 
> Given we'll need a follow-up PR here anyways once TE common supports other compute dtypes, I feel like it's easier to have this PR just be the bugfix and wait for the generalizing until TE common changes have merged.
> 
> Wdyt?

That makes sense. I did not think about the possibility that users could pass in a different dtype and mess the backend kernels up as it is not ready for anything but fp32. It will require another PR anyways. I will keep this as fp32 and add the `NVTE_CHECK`

File: transformer_engine/jax/csrc/extensions/router.cpp [hunk 2]
```diff
@@ -97,17 +103,23 @@ Error_Type FusedTopkWithScoreFunctionBackwardFFI(
     Result_Type grad_logits_buf,  // [num_tokens, num_experts]
     int64_t topk, int64_t use_pre_softmax, double scaling_factor,
     JAXX_Score_Function score_function, int64_t compute_aux_scores) {
-  auto dtype = convert_ffi_datatype_to_te_dtype(intermediate_buf.element_type());
+  // intermediate uses CompType (set by the abstract via ROUTER_COMP_DTYPE).
+  NVTE_CHECK(convert_ffi_datatype_to_te_dtype(intermediate_buf.element_type()) == DType::kFloat32,
+             "intermediate_output must be float32 (CompType); got dtype ",
+             static_cast<int>(convert_ffi_datatype_to_te_dtype(intermediate_buf.element_type())),
+             ". Check ROUTER_COMP_DTYPE in cpp_extensions/router.py.");
```
[the remark this answers is not in the document]
Comment (phu0ngng) [reply transformer_engine/jax/csrc/extensions/router.cpp]: instead of static_cast, how about using `to_string`?
